diff --git a/sourcekit-lsp/Utilities/build-script-helper.py b/sourcekit-lsp/Utilities/build-script-helper.py
index 88141cdd..d264f10c 100755
--- a/sourcekit-lsp/Utilities/build-script-helper.py
+++ b/sourcekit-lsp/Utilities/build-script-helper.py
@@ -136,9 +136,6 @@ def get_swiftpm_options(swift_exec: str, args: argparse.Namespace, suppress_verb
     if '-android' in build_target:
         swiftpm_args += [
             '-Xlinker', '-rpath', '-Xlinker', '$ORIGIN/../lib/swift/android',
+            '-Xlinker', '-landroid-spawn',
-            # SwiftPM will otherwise try to compile against GNU strerror_r on
-            '-Xlinker', '-landroid-spawn',
-            # Android and fail.
-            '-Xswiftc', '-Xcc', '-Xswiftc', '-U_GNU_SOURCE',
         ]
     elif not build_os.startswith('macosx'):
         # Library rpath for swift, dispatch, Foundation, etc. when installing
diff --git a/swiftpm/Utilities/bootstrap b/swiftpm/Utilities/bootstrap
index 156bf002a..d891da556 100755
--- a/swiftpm/Utilities/bootstrap
+++ b/swiftpm/Utilities/bootstrap
@@ -934,10 +934,7 @@ def get_swiftpm_flags(args):
         build_flags.extend(["-Xcc", "-I/usr/local/include"])
         build_flags.extend(["-Xlinker", "-L/usr/local/lib"])
 
-    # Don't use GNU strerror_r on Android.
-    if '-android' in args.build_target:
-        build_flags.extend(["-Xswiftc", "-Xcc", "-Xswiftc", "-U_GNU_SOURCE"])
-        build_flags.extend(["-Xlinker", "-landroid-spawn"])
+    build_flags.extend(["-Xlinker", "-landroid-spawn"])
 
     cross_compile_hosts = args.cross_compile_hosts
     if cross_compile_hosts:
