diff --git a/swift/utils/swift_build_support/swift_build_support/targets.py b/swift/utils/swift_build_support/swift_build_support/targets.py
index 9932b854cb6..ad3ac757665 100644
--- a/swift/utils/swift_build_support/swift_build_support/targets.py
+++ b/swift/utils/swift_build_support/swift_build_support/targets.py
@@ -167,9 +167,10 @@ class AndroidPlatform(Platform):
         options = cmake.CMakeOptions()
         options.define('CMAKE_SYSTEM_NAME', 'Android')
         options.define('CMAKE_SYSTEM_VERSION' , args.android_api_level)
-        options.define('CMAKE_SYSTEM_PROCESSOR', args.android_arch if not
-                       args.android_arch == 'armv7'
-                       else 'armv7-a')
+        arm_arch_to_abi = { 'armv7' : 'armeabi-v7a', 'aarch64' : 'arm64-v8a' }
+        options.define('CMAKE_ANDROID_ARCH_ABI', args.android_arch if not
+                       args.android_arch.startswith('a')
+                       else arm_arch_to_abi[args.android_arch])
         options.define('CMAKE_ANDROID_NDK:PATH', args.android_ndk)
         return options
 
diff --git a/swift-corelibs-foundation/Sources/Foundation/CMakeLists.txt b/swift-corelibs-foundation/Sources/Foundation/CMakeLists.txt
index 016bf294..5c42986a 100644
--- a/swift-corelibs-foundation/Sources/Foundation/CMakeLists.txt
+++ b/swift-corelibs-foundation/Sources/Foundation/CMakeLists.txt
@@ -167,6 +167,12 @@ if(CMAKE_SYSTEM_NAME STREQUAL Windows)
     $<TARGET_OBJECTS:CoreFoundationResources>)
 elseif(NOT CMAKE_SYSTEM_NAME STREQUAL Darwin)
   target_link_options(Foundation PRIVATE "SHELL:-no-toolchain-stdlib-rpath")
+  if(${CMAKE_SYSTEM_NAME} STREQUAL Android)
+    target_link_libraries(Foundation PRIVATE android-spawn)
+    list(GET CMAKE_FIND_ROOT_PATH 0 SPAWN_DIR)
+    target_include_directories(Foundation PUBLIC ${SPAWN_DIR}/usr/include)
+    target_link_directories(Foundation PUBLIC ${SPAWN_DIR}/usr/lib)
+  endif()
 endif()
 
 
