diff --git a/Sources/NIOSSL/SSLContext.swift b/Sources/NIOSSL/SSLContext.swift
index 8b82976..3abecab 100644
--- a/Sources/NIOSSL/SSLContext.swift
+++ b/Sources/NIOSSL/SSLContext.swift
@@ -853,7 +853,7 @@ internal class DirectoryContents: Sequence, IteratorProtocol {
     
     init(path: String) {
         self.path = path
-        self.dir = opendir(path)
+        self.dir = opendir(path)!
     }
     
     func next() -> String? {
