diff --git a/sourcekit-lsp/Utilities/build-script-helper.py b/sourcekit-lsp/Utilities/build-script-helper.py
index a64795c..bc334a8 100755
--- a/sourcekit-lsp/Utilities/build-script-helper.py
+++ b/sourcekit-lsp/Utilities/build-script-helper.py
@@ -66,7 +66,7 @@ def get_swiftpm_options(args):
     ]
 
   if args.cross_compile_host:
-    swiftpm_args += ['--destination', args.cross_compile_config]
+    swiftpm_args += ['--destination', args.cross_compile_config, '-Xlinker', '-landroid-spawn']
 
   return swiftpm_args
 
