diff --git a/swift/utils/build-script-impl b/swift/utils/build-script-impl
index 16e05052609..7ab8cebfab8 100755
--- a/swift/utils/build-script-impl
+++ b/swift/utils/build-script-impl
@@ -2622,6 +2622,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     echo "Cleaning the libdispatch build directory"
                     call rm -rf "${LIBDISPATCH_BUILD_DIR}"
                   fi
+                  call ln -sf "${SWIFT_BUILD_PATH}/lib/swift" "${ANDROID_NDK}/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/lib"
 
                   cmake_options=(
                     -DENABLE_SWIFT=YES
diff --git a/swift-corelibs-foundation/Package.swift b/swift-corelibs-foundation/Package.swift
index 2b5b5e05..60f1dde8 100644
--- a/swift-corelibs-foundation/Package.swift
+++ b/swift-corelibs-foundation/Package.swift
@@ -245,7 +245,8 @@ let package = Package(
                 "BlockRuntime",
                 "CMakeLists.txt"
             ],
-            cSettings: coreFoundationBuildSettings
+            cSettings: coreFoundationBuildSettings,
+            linkerSettings: [.linkedLibrary("log", .when(platforms: [.android]))]
         ),
         .target(
             name: "BlocksRuntime",
@@ -262,7 +263,7 @@ let package = Package(
             name: "_CFXMLInterface",
             dependencies: [
                 "CoreFoundation",
-                .target(name: "Clibxml2", condition: .when(platforms: [.linux])),
+                .target(name: "Clibxml2", condition: .when(platforms: [.linux, .android])),
             ],
             path: "Sources/_CFXMLInterface",
             exclude: [
@@ -275,7 +276,7 @@ let package = Package(
             name: "_CFURLSessionInterface",
             dependencies: [
                 "CoreFoundation",
-                .target(name: "Clibcurl", condition: .when(platforms: [.linux])),
+                .target(name: "Clibcurl", condition: .when(platforms: [.linux, .android])),
             ],
             path: "Sources/_CFURLSessionInterface",
             exclude: [
@@ -348,7 +349,7 @@ let package = Package(
                 "FoundationNetworking",
                 "XCTest",
                 "Testing",
-                .target(name: "xdgTestHelper", condition: .when(platforms: [.linux]))
+                .target(name: "xdgTestHelper", condition: .when(platforms: [.linux, .android]))
             ],
             resources: [
                 .copy("Foundation/Resources")
diff --git a/swift-corelibs-foundation/Sources/Foundation/NSPathUtilities.swift b/swift-corelibs-foundation/Sources/Foundation/NSPathUtilities.swift
index 8043ef00..23b602f6 100644
--- a/swift-corelibs-foundation/Sources/Foundation/NSPathUtilities.swift
+++ b/swift-corelibs-foundation/Sources/Foundation/NSPathUtilities.swift
@@ -10,6 +10,8 @@
 @_implementationOnly import CoreFoundation
 #if os(Windows)
 import WinSDK
+#elseif canImport(Android)
+import Android
 #elseif os(WASI)
 import WASILibc
 #endif
diff --git a/swift-corelibs-foundation/Sources/Foundation/Process.swift b/swift-corelibs-foundation/Sources/Foundation/Process.swift
index 758dd1df..02970992 100644
--- a/swift-corelibs-foundation/Sources/Foundation/Process.swift
+++ b/swift-corelibs-foundation/Sources/Foundation/Process.swift
@@ -927,8 +927,8 @@ open class Process: NSObject, @unchecked Sendable {
         }
         let useFallbackChdir: Bool
         if let dir = currentDirectoryURL?.path {
-            let chdirResult = _CFPosixSpawnFileActionsChdir(fileActions, dir)
-            useFallbackChdir = chdirResult == ENOSYS
+            // let chdirResult = _CFPosixSpawnFileActionsChdir(fileActions, dir)
+            useFallbackChdir = true ; let chdirResult = ENOSYS
             if !useFallbackChdir {
                 try _throwIfPosixError(chdirResult)
             }
diff --git a/swift-corelibs-foundation/Sources/Testing/Testing.swift b/swift-corelibs-foundation/Sources/Testing/Testing.swift
index 712d9dee..2483c14e 100644
--- a/swift-corelibs-foundation/Sources/Testing/Testing.swift
+++ b/swift-corelibs-foundation/Sources/Testing/Testing.swift
@@ -11,6 +11,8 @@
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #elseif os(WASI)
 import WASILibc
 #elseif canImport(CRT)
From dfc3d9dba1e8f0093851832716e9d050447e1aac
From: Alex Lorenz <arphaman@gmail.com>
Date: Wed, 23 Oct 2024 13:06:05 -0700
Subject: [PATCH] [android] fix the android build

Platform.swift regressed after 71eefee7fca08ecbeacc89098b74ab80319baa6b
---
 Sources/FoundationEssentials/Platform.swift | 35 ++++++++++++++++-----
 1 file changed, 27 insertions(+), 8 deletions(-)

diff --git a/swift-foundation/Sources/FoundationEssentials/Platform.swift b/swift-foundation/Sources/FoundationEssentials/Platform.swift
index d8b8fe6f..04875b52 100644
--- a/swift-foundation/Sources/FoundationEssentials/Platform.swift
+++ b/swift-foundation/Sources/FoundationEssentials/Platform.swift
@@ -29,8 +29,7 @@ fileprivate let _pageSize: Int = {
 // WebAssembly defines a fixed page size
 fileprivate let _pageSize: Int = 65_536
 #elseif canImport(Android)
-import Bionic
-import unistd
+import Android
 fileprivate let _pageSize: Int = Int(getpagesize())
 #elseif canImport(Glibc)
 import Glibc
@@ -142,7 +141,7 @@ extension Platform {
     typealias Operation<Input, Output> = (Input, UnsafeMutablePointer<Output>, UnsafeMutablePointer<CChar>, Int, UnsafeMutablePointer<UnsafeMutablePointer<Output>?>) -> Int32
     #endif
     
-    private static func withUserGroupBuffer<Input, Output, R>(_ input: Input, _ output: Output, sizeProperty: Int32, operation: Operation<Input, Output>, block: (Output) throws -> R) rethrows -> R? {
+    private static func withUserGroupBuffer<Input, Output, R>(_ input: Input, _ output: Output, sizeProperty: Int32, operation: Operation<Input, Output>, block: (Output) throws -> R?) rethrows -> R? {
         var bufferLen = sysconf(sizeProperty)
         if bufferLen == -1 {
             bufferLen = 4096 // Generous default size estimate
@@ -172,31 +171,51 @@ extension Platform {
     
     static func name(forUID uid: uid_t) -> String? {
         withUserGroupBuffer(uid, passwd(), sizeProperty: Int32(_SC_GETPW_R_SIZE_MAX), operation: getpwuid_r) {
-            String(cString: $0.pw_name)
+            // Android's pw_name `char *`` is nullable, so always coerce to a nullable pointer
+            // in order to be compatible with Android.
+            let pw_name: UnsafeMutablePointer<CChar>? = $0.pw_name
+            return pw_name.flatMap { String(cString: $0) }
         }
     }
     
     static func fullName(forUID uid: uid_t) -> String? {
         withUserGroupBuffer(uid, passwd(), sizeProperty: Int32(_SC_GETPW_R_SIZE_MAX), operation: getpwuid_r) {
-            String(cString: $0.pw_gecos)
+#if os(Android) && (arch(i386) || arch(arm))
+            // pw_gecos isn't available on 32-bit Android.
+            let pw_gecos: UnsafeMutablePointer<CChar>? = nil
+#else
+            // Android's pw_gecos `char *`` is nullable, so always coerce to a nullable pointer
+            // in order to be compatible with Android.
+            let pw_gecos: UnsafeMutablePointer<CChar>? = $0.pw_gecos
+#endif
+            return pw_gecos.flatMap { String(cString: $0) }
         }
     }
     
     static func name(forGID gid: gid_t) -> String? {
         withUserGroupBuffer(gid, group(), sizeProperty: Int32(_SC_GETGR_R_SIZE_MAX), operation: getgrgid_r) {
-            String(cString: $0.gr_name)
+            // Android's gr_name `char *`` is nullable, so always coerce to a nullable pointer
+            // in order to be compatible with Android.
+            let gr_name: UnsafeMutablePointer<CChar>? = $0.gr_name
+            return gr_name.flatMap { String(cString: $0) }
         }
     }
     
     static func homeDirectory(forUserName userName: String) -> String? {
         withUserGroupBuffer(userName, passwd(), sizeProperty: Int32(_SC_GETPW_R_SIZE_MAX), operation: getpwnam_r) {
-            String(cString: $0.pw_dir)
+            // Android's pw_dir `char *`` is nullable, so always coerce to a nullable pointer
+            // in order to be compatible with Android.
+            let pw_dir: UnsafeMutablePointer<CChar>? = $0.pw_dir
+            return pw_dir.flatMap { String(cString: $0) }
         }
     }
     
     static func homeDirectory(forUID uid: uid_t) -> String? {
         withUserGroupBuffer(uid, passwd(), sizeProperty: Int32(_SC_GETPW_R_SIZE_MAX), operation: getpwuid_r) {
-            String(cString: $0.pw_dir)
+            // Android's pw_dir `char *`` is nullable, so always coerce to a nullable pointer
+            // in order to be compatible with Android.
+            let pw_dir: UnsafeMutablePointer<CChar>? = $0.pw_dir
+            return pw_dir.flatMap { String(cString: $0) }
         }
     }
 }
