diff --git a/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift b/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift
index 0be22ce8..33ca7ee5 100644
--- a/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift
+++ b/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/JSONRPCConnection.swift
@@ -16,6 +16,9 @@ public import Foundation
 public import LanguageServerProtocol
 import SKLogging
 import SwiftExtensions
+#if canImport(Android)
+import Android
+#endif
 #else
 import Dispatch
 import Foundation
