diff --git a/Sources/NIOSSHClient/ExecHandler.swift b/Sources/NIOSSHClient/ExecHandler.swift
index 8dd244e..9f1e5d5 100644
--- a/Sources/NIOSSHClient/ExecHandler.swift
+++ b/Sources/NIOSSHClient/ExecHandler.swift
@@ -100,7 +100,7 @@ final class ExampleExecHandler: ChannelDuplexHandler {
         case .stdErr:
             // We just write to stderr directly, pipe channel can't help us here.
             bytes.withUnsafeReadableBytes { str in
-                let rc = fwrite(str.baseAddress, 1, str.count, stderr)
+                let rc = fwrite(str.baseAddress!, 1, str.count, stderr)
                 precondition(rc == str.count)
             }
 
