diff --git a/swift/utils/build-script-impl b/swift/utils/build-script-impl
index 890fc5e1419..a9bdd7f8842 100755
--- a/swift/utils/build-script-impl
+++ b/swift/utils/build-script-impl
@@ -2331,7 +2331,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     continue
                 fi
 
-                if [[ "${BUILD_SWIFT_TOOLS}" == "0" ]]; then
+                if [[ "${CROSS_COMPILE_BUILD_SWIFT_TOOLS}" == "0" ]]; then
                     echo "Skipping building Foundation Macros for ${host}, because the host tools are not being built"
                     continue
                 fi
@@ -2920,7 +2920,7 @@ for host in "${ALL_HOSTS[@]}"; do
                     continue
                 fi
 
-                if [[ "${BUILD_SWIFT_TOOLS}" == "0" && "${product}" == "foundation_macros" ]]; then
+                if [[ "${CROSS_COMPILE_BUILD_SWIFT_TOOLS}" == "0" && "${product}" == "foundation_macros" ]]; then
                     echo "Skipping installing Foundation Macros for ${host}, because the host tools are not being built"
                     continue
                 fi
diff --git a/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake b/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake
index 3da519e..e2dd8ed 100644
--- a/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake
+++ b/swift-corelibs-libdispatch/cmake/modules/SwiftSupport.cmake
@@ -4,6 +4,10 @@ if(NOT dispatch_MODULE_TRIPLE)
   set(module_triple_command "${CMAKE_Swift_COMPILER}" -print-target-info)
   if(CMAKE_Swift_COMPILER_TARGET)
     list(APPEND module_triple_command -target ${CMAKE_Swift_COMPILER_TARGET})
+  else()
+    set(arg_list ${CMAKE_Swift_FLAGS})
+    separate_arguments(arg_list)
+    list(APPEND module_triple_command ${arg_list})
   endif()
   execute_process(COMMAND ${module_triple_command} OUTPUT_VARIABLE target_info_json)
 
diff --git a/swift-corelibs-libdispatch/src/swift/CMakeLists.txt b/swift-corelibs-libdispatch/src/swift/CMakeLists.txt
index 38bef37..d0ddf98 100644
--- a/swift-corelibs-libdispatch/src/swift/CMakeLists.txt
+++ b/swift-corelibs-libdispatch/src/swift/CMakeLists.txt
@@ -15,6 +15,7 @@ add_library(swiftDispatch
   Source.swift
   Time.swift
   Wrapper.swift)
+set(CMAKE_SHARED_LINKER_FLAGS "")
 target_compile_options(swiftDispatch PRIVATE
   "SHELL:-Xcc -fblocks"
   "SHELL:-Xcc -fmodule-map-file=${PROJECT_SOURCE_DIR}/dispatch/module.modulemap"
