diff --git a/swift/stdlib/public/SwiftShims/swift/shims/LibcShims.h b/swift/stdlib/public/SwiftShims/swift/shims/LibcShims.h
index 6a34bc4746a..bc4a02a6eed 100644
--- a/swift/stdlib/public/SwiftShims/swift/shims/LibcShims.h
+++ b/swift/stdlib/public/SwiftShims/swift/shims/LibcShims.h
@@ -61,7 +61,7 @@ SWIFT_READONLY
 static inline int _swift_stdlib_memcmp(const void *s1, const void *s2,
                                        __swift_size_t n) {
 // FIXME: Is there a way to identify Glibc specifically?
-#if defined(__gnu_linux__) || defined(__ANDROID__) || defined(__musl__)
+#if defined(__gnu_linux__) || defined(__musl__)
   extern int memcmp(const void * _Nonnull, const void * _Nonnull, __swift_size_t);
 #else
   extern int memcmp(const void * _Null_unspecified, const void * _Null_unspecified, __swift_size_t);
