diff --git a/sourcekit-lsp/Utilities/build-script-helper.py b/sourcekit-lsp/Utilities/build-script-helper.py
index 30cdb14..d68622a 100755
--- a/sourcekit-lsp/Utilities/build-script-helper.py
+++ b/sourcekit-lsp/Utilities/build-script-helper.py
@@ -92,7 +92,7 @@ def get_swiftpm_options(swift_exec, args):
       swiftpm_args += ["--arch", "x86_64", "--arch", "arm64"]
     elif re.match('android-', args.cross_compile_host):
       print('Cross-compiling for %s' % args.cross_compile_host)
-      swiftpm_args += ['--destination', args.cross_compile_config]
+      swiftpm_args += ['--destination', args.cross_compile_config, '-Xlinker', '-landroid-spawn']
     else:
       error("cannot cross-compile for %s" % args.cross_compile_host)
 
