diff --git a/sourcekit-lsp/Utilities/build-script-helper.py b/sourcekit-lsp/Utilities/build-script-helper.py
index a64795c..bc334a8 100755
--- a/sourcekit-lsp/Utilities/build-script-helper.py
+++ b/sourcekit-lsp/Utilities/build-script-helper.py
@@ -114,7 +114,7 @@ def handle_invocation(swift_exec, args):
     env['TSAN_OPTIONS'] = 'halt_on_error=true'
 
   print('Cleaning ' + args.build_path)
-  shutil.rmtree(args.build_path, ignore_errors=True)
+#  shutil.rmtree(args.build_path, ignore_errors=True)
 
   env['SWIFT_EXEC'] = '%sc' % (swift_exec)
 
