diff --git a/swift/stdlib/public/runtime/CrashHandlerLinux.cpp b/swift/stdlib/public/runtime/CrashHandlerLinux.cpp
index e737a64954c..c874433bf6f 100644
--- a/swift/stdlib/public/runtime/CrashHandlerLinux.cpp
+++ b/swift/stdlib/public/runtime/CrashHandlerLinux.cpp
@@ -660,9 +660,11 @@ memserver_fault(int sig) {
 ssize_t __attribute__((noinline))
 memserver_read(void *to, const void *from, size_t len) {
   if (memserver_has_ptrace) {
+#if defined(__ANDROID_API__) && __ANDROID_API__ >= 23
     struct iovec local = { to, len };
     struct iovec remote = { const_cast<void *>(from), len };
     return process_vm_readv(memserver_pid, &local, 1, &remote, 1, 0);
+#endif
   } else {
     if (!sigsetjmp(memserver_fault_buf, 1)) {
       memcpy(to, from, len);
@@ -682,7 +684,11 @@ memserver_entry(void *dummy __attribute__((unused))) {
   prctl(PR_SET_NAME, "[backtrace]");
 #endif
 
+#if defined(__ANDROID_API__) && __ANDROID_API__ < 23
+  memserver_has_ptrace = false;
+#else
   memserver_has_ptrace = !!prctl(PR_CAPBSET_READ, CAP_SYS_PTRACE);
+#endif
 
   if (!memserver_has_ptrace) {
     struct sigaction sa;
