diff --git a/Sources/_CryptoExtras/Util/BoringSSLHelpers.swift b/Sources/_CryptoExtras/Util/BoringSSLHelpers.swift
index b306c7f..0ea88a7 100644
--- a/Sources/_CryptoExtras/Util/BoringSSLHelpers.swift
+++ b/Sources/_CryptoExtras/Util/BoringSSLHelpers.swift
@@ -86,7 +86,11 @@ extension FixedWidthInteger {
             CCryptoBoringSSL_BN_clear(&bn)
         }
 
+        #if arch(arm)
+        CCryptoBoringSSL_BN_set_word(&bn, UInt32(self))
+        #else
         CCryptoBoringSSL_BN_set_word(&bn, UInt64(self))
+        #endif
 
         return try block(&bn)
     }
