diff --git a/Sources/NIOSSHClient/InteractivePasswordPromptDelegate.swift b/Sources/NIOSSHClient/InteractivePasswordPromptDelegate.swift
index f615cc1..c0b6143 100644
--- a/Sources/NIOSSHClient/InteractivePasswordPromptDelegate.swift
+++ b/Sources/NIOSSHClient/InteractivePasswordPromptDelegate.swift
@@ -45,7 +45,7 @@ final class InteractivePasswordPromptDelegate: NIOSSHClientUserAuthenticationDel
             }
 
             if self.password == nil {
-                #if os(Windows)
+                #if os(Windows) || os(Android)
                 print("Password: ", terminator: "")
                 self.password = readLine() ?? ""
                 #else
