diff --git a/Sources/NIOSSL/SSLContext.swift b/Sources/NIOSSL/SSLContext.swift
index 01dba6d..d658c0e 100644
--- a/Sources/NIOSSL/SSLContext.swift
+++ b/Sources/NIOSSL/SSLContext.swift
@@ -41,7 +41,7 @@ internal enum FileSystemObject {
             return nil
         }
 
-#if os(Android)
+#if os(Android) && arch(arm)
         return (statObj.st_mode & UInt32(Glibc.S_IFDIR)) != 0 ? .directory : .file
 #else
         return (statObj.st_mode & S_IFDIR) != 0 ? .directory : .file
@@ -595,7 +595,11 @@ extension NIOSSLContext {
         var buffer = stat()
         let _ = try Posix.lstat(path: path, buf: &buffer)
         // Check the mode to make sure this is a symlink
+#if os(Android) && arch(arm)
+        if (buffer.st_mode & UInt32(Glibc.S_IFMT)) != UInt32(Glibc.S_IFLNK) { return false }
+#else
         if (buffer.st_mode & S_IFMT) != S_IFLNK { return false }
+#endif
 
         // Return true at this point because the file format is considered to be in rehash format and a symlink.
         // Rehash format being "%08lx.%d" or HHHHHHHH.D
