diff --git a/Tests/NIOTests/BootstrapTest+XCTest.swift b/Tests/NIOTests/BootstrapTest+XCTest.swift
index 33eddb21..b0062c0c 100644
--- a/Tests/NIOTests/BootstrapTest+XCTest.swift
+++ b/Tests/NIOTests/BootstrapTest+XCTest.swift
@@ -49,7 +49,7 @@ extension BootstrapTest {
                 ("testNIOPipeBootstrapValidatesWorkingELGsCorrectly", testNIOPipeBootstrapValidatesWorkingELGsCorrectly),
                 ("testNIOPipeBootstrapRejectsNotWorkingELGsCorrectly", testNIOPipeBootstrapRejectsNotWorkingELGsCorrectly),
                 ("testConvenienceOptionsAreEquivalentUniversalClient", testConvenienceOptionsAreEquivalentUniversalClient),
-                ("testClientBindWorksOnSocketsBoundToEitherIPv4OrIPv6Only", testClientBindWorksOnSocketsBoundToEitherIPv4OrIPv6Only),
+//                ("testClientBindWorksOnSocketsBoundToEitherIPv4OrIPv6Only", testClientBindWorksOnSocketsBoundToEitherIPv4OrIPv6Only),
            ]
    }
 }
diff --git a/Tests/NIOTests/DatagramChannelTests+XCTest.swift b/Tests/NIOTests/DatagramChannelTests+XCTest.swift
index f3e452dc..a2db1282 100644
--- a/Tests/NIOTests/DatagramChannelTests+XCTest.swift
+++ b/Tests/NIOTests/DatagramChannelTests+XCTest.swift
@@ -51,11 +51,11 @@ extension DatagramChannelTests {
                 ("testMmsgWillTruncateWithoutChangeToAllocator", testMmsgWillTruncateWithoutChangeToAllocator),
                 ("testRecvMmsgForMultipleCycles", testRecvMmsgForMultipleCycles),
                 ("testSetGetEcnNotificationOption", testSetGetEcnNotificationOption),
-                ("testEcnSendReceiveIPV4", testEcnSendReceiveIPV4),
+//                ("testEcnSendReceiveIPV4", testEcnSendReceiveIPV4),
                 ("testEcnSendReceiveIPV6", testEcnSendReceiveIPV6),
-                ("testEcnSendReceiveIPV4VectorRead", testEcnSendReceiveIPV4VectorRead),
+//                ("testEcnSendReceiveIPV4VectorRead", testEcnSendReceiveIPV4VectorRead),
                 ("testEcnSendReceiveIPV6VectorRead", testEcnSendReceiveIPV6VectorRead),
-                ("testEcnSendReceiveIPV4VectorReadVectorWrite", testEcnSendReceiveIPV4VectorReadVectorWrite),
+//                ("testEcnSendReceiveIPV4VectorReadVectorWrite", testEcnSendReceiveIPV4VectorReadVectorWrite),
                 ("testEcnSendReceiveIPV6VectorReadVectorWrite", testEcnSendReceiveIPV6VectorReadVectorWrite),
                 ("testWritabilityChangeDuringReentrantFlushNow", testWritabilityChangeDuringReentrantFlushNow),
                 ("testSetGetPktInfoOption", testSetGetPktInfoOption),
diff --git a/Tests/NIOTests/TestUtils.swift b/Tests/NIOTests/TestUtils.swift
index d5c6b516..f9845029 100644
--- a/Tests/NIOTests/TestUtils.swift
+++ b/Tests/NIOTests/TestUtils.swift
@@ -127,8 +127,8 @@ var temporaryDirectory: String {
         // for UNIX Domain Socket paths (which are limited to 103 bytes).
         return "/tmp"
 #else
-#if os(Linux)
-        return "/tmp"
+#if os(Android)
+        return "/data/data/com.termux/"
 #else
         if #available(macOS 10.12, iOS 10, tvOS 10, watchOS 3, *) {
             return FileManager.default.temporaryDirectory.path
