diff --git a/sourcekit-lsp/Utilities/build-script-helper.py b/sourcekit-lsp/Utilities/build-script-helper.py
index 17ad9e3..953fffd 100755
--- a/sourcekit-lsp/Utilities/build-script-helper.py
+++ b/sourcekit-lsp/Utilities/build-script-helper.py
@@ -179,6 +179,9 @@ def get_swiftpm_environment_variables(swift_exec: str, args: argparse.Namespace)
     if args.action == 'test' and not args.skip_long_tests:
         env['SOURCEKIT_LSP_ENABLE_LONG_TESTS'] = '1'
 
+    if args.cross_compile_host and re.match('android-', args.cross_compile_host):
+        env['PKG_CONFIG_PATH'] = os.path.join(os.path.split(args.cross_compile_config)[0], 'lib/pkgconfig')
+
     return env
 
 
diff --git a/swift/utils/swift_build_support/swift_build_support/products/swiftpm.py b/swift/utils/swift_build_support/swift_build_support/products/swiftpm.py
index 42a2ff914b4..3ea91589a57 100644
--- a/swift/utils/swift_build_support/swift_build_support/products/swiftpm.py
+++ b/swift/utils/swift_build_support/swift_build_support/products/swiftpm.py
@@ -67,7 +67,7 @@ class SwiftPM(product.Product):
 
         helper_cmd += [
             "--swiftc-path", swiftc,
-            "--clang-path", self.toolchain.cc,
+            "--clang-path", os.path.join(toolchain_path, "bin", "clang"),
             "--cmake-path", self.toolchain.cmake,
             "--ninja-path", self.toolchain.ninja,
             "--build-dir", self.build_dir,
diff --git a/swiftpm/Utilities/bootstrap b/swiftpm/Utilities/bootstrap
index 085056de..1ded1a90 100755
--- a/swiftpm/Utilities/bootstrap
+++ b/swiftpm/Utilities/bootstrap
@@ -615,6 +615,9 @@
         "CC=" + args.clang_path
     ]
 
+    if args.cross_compile_hosts and re.match('android-', args.cross_compile_hosts):
+        swiftpm_args.append("PKG_CONFIG_PATH=" + os.path.join(os.path.split(args.cross_compile_config)[0], "lib/pkgconfig"))
+
     if args.bootstrap:
         note("Building SwiftPM (with a freshly built swift-bootstrap)")
         swiftpm_args.append("SWIFTPM_CUSTOM_LIBS_DIR=" + os.path.join(args.bootstrap_dir, "pm"))
