diff --git a/Tests/LinuxMain.swift b/Tests/LinuxMain.swift
index d2d2f05..f259827 100644
--- a/Tests/LinuxMain.swift
+++ b/Tests/LinuxMain.swift
@@ -29,10 +29,9 @@ import XCTest
 // without the compiler noticing that we're calling a deprecated function.
 // This hack exists so we can deprecate individual tests which test deprecated functionality without
 // getting a compiler warning...
-protocol LinuxMainRunner { func run() }
-class LinuxMainRunnerImpl: LinuxMainRunner {
+@main class LinuxMainRunnerImpl {
    @available(*, deprecated, message: "not actually deprecated. Just deprecated to allow deprecated tests (which test deprecated functionality) without warnings")
-   func run() {
+   static func main() {
        XCTMain([
              testCase(ByteBufferBIOTest.allTests),
              testCase(CertificateVerificationTests.allTests),
@@ -53,5 +52,4 @@ class LinuxMainRunnerImpl: LinuxMainRunner {
         ])
     }
 }
-(LinuxMainRunnerImpl() as LinuxMainRunner).run()
 #endif
diff --git a/Tests/NIOSSLTests/NIOSSLIntegrationTest+XCTest.swift b/Tests/NIOSSLTests/NIOSSLIntegrationTest+XCTest.swift
index 51dd4a6..25e15d6 100644
--- a/Tests/NIOSSLTests/NIOSSLIntegrationTest+XCTest.swift
+++ b/Tests/NIOSSLTests/NIOSSLIntegrationTest+XCTest.swift
@@ -69,7 +69,7 @@ extension NIOSSLIntegrationTest {
                 ("testWriteFromFailureOfWrite", testWriteFromFailureOfWrite),
                 ("testChannelInactiveDuringHandshakeSucceeded", testChannelInactiveDuringHandshakeSucceeded),
                 ("testTrustedFirst", testTrustedFirst),
-                ("testGiantWrite", testGiantWrite),
+                //("testGiantWrite", testGiantWrite),
            ]
    }
 }
