diff --git a/llbuild/products/llbuildSwift/BuildDBBindings.swift b/llbuild/products/llbuildSwift/BuildDBBindings.swift
index db8a9b72..78652cc7 100644
--- a/llbuild/products/llbuildSwift/BuildDBBindings.swift
+++ b/llbuild/products/llbuildSwift/BuildDBBindings.swift
@@ -17,6 +17,8 @@ import WinSDK
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #else
 #error("Missing libc or equivalent")
 #endif
diff --git a/llbuild/products/llbuildSwift/BuildKey.swift b/llbuild/products/llbuildSwift/BuildKey.swift
index 1d89c534..a59e895a 100644
--- a/llbuild/products/llbuildSwift/BuildKey.swift
+++ b/llbuild/products/llbuildSwift/BuildKey.swift
@@ -17,6 +17,8 @@ import WinSDK
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #else
 #error("Missing libc or equivalent")
 #endif
diff --git a/llbuild/products/llbuildSwift/BuildSystemBindings.swift b/llbuild/products/llbuildSwift/BuildSystemBindings.swift
index 85c8fe24..9566f66b 100644
--- a/llbuild/products/llbuildSwift/BuildSystemBindings.swift
+++ b/llbuild/products/llbuildSwift/BuildSystemBindings.swift
@@ -17,6 +17,8 @@ import WinSDK
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #else
 #error("Missing libc or equivalent")
 #endif
@@ -1235,7 +1237,7 @@ public final class BuildSystem {
         #elseif os(Windows)
         info.pointee.mod_time.seconds = UInt64(s.st_mtime)
         info.pointee.mod_time.nanoseconds = 0
-        #elseif canImport(Glibc) || canImport(Musl)
+        #elseif canImport(Glibc) || canImport(Musl) || canImport(Bionic)
         info.pointee.mod_time.seconds = UInt64(s.st_mtim.tv_sec)
         info.pointee.mod_time.nanoseconds = UInt64(s.st_mtim.tv_nsec)
         #else
diff --git a/llbuild/products/llbuildSwift/BuildValue.swift b/llbuild/products/llbuildSwift/BuildValue.swift
index 2caa5947..b2af480f 100644
--- a/llbuild/products/llbuildSwift/BuildValue.swift
+++ b/llbuild/products/llbuildSwift/BuildValue.swift
@@ -17,6 +17,8 @@ import WinSDK
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #else
 #error("Missing libc or equivalent")
 #endif
diff --git a/llbuild/products/llbuildSwift/Internals.swift b/llbuild/products/llbuildSwift/Internals.swift
index 4e134288..3f5c4fe4 100644
--- a/llbuild/products/llbuildSwift/Internals.swift
+++ b/llbuild/products/llbuildSwift/Internals.swift
@@ -17,6 +17,8 @@ import WinSDK
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #else
 #error("Missing libc or equivalent")
 #endif
diff --git a/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/DisableSigpipe.swift b/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/DisableSigpipe.swift
index e9967f64..0630fe53 100644
--- a/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/DisableSigpipe.swift
+++ b/sourcekit-lsp/Sources/LanguageServerProtocolJSONRPC/DisableSigpipe.swift
@@ -14,9 +14,11 @@
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Android)
+import Android
 #endif
 
-#if canImport(Glibc) || canImport(Musl)
+#if canImport(Glibc) || canImport(Musl) || canImport(Android)
 // This is a lazily initialised global variable that when read for the first time, will ignore SIGPIPE.
 private let globallyIgnoredSIGPIPE: Bool = {
   /* no F_SETNOSIGPIPE on Linux :( */
diff --git a/sourcekit-lsp/Sources/SourceKitD/SKDRequestArray.swift b/sourcekit-lsp/Sources/SourceKitD/SKDRequestArray.swift
index a68c4ae3..f8aed9d7 100644
--- a/sourcekit-lsp/Sources/SourceKitD/SKDRequestArray.swift
+++ b/sourcekit-lsp/Sources/SourceKitD/SKDRequestArray.swift
@@ -18,6 +18,8 @@ import Glibc
 import Musl
 #elseif canImport(CRT)
 import CRT
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 extension SourceKitD {
diff --git a/sourcekit-lsp/Sources/SourceKitD/SKDRequestDictionary.swift b/sourcekit-lsp/Sources/SourceKitD/SKDRequestDictionary.swift
index 99e02502..7ff965bc 100644
--- a/sourcekit-lsp/Sources/SourceKitD/SKDRequestDictionary.swift
+++ b/sourcekit-lsp/Sources/SourceKitD/SKDRequestDictionary.swift
@@ -19,6 +19,8 @@ import Glibc
 import Musl
 #elseif canImport(CRT)
 import CRT
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 /// Values that can be stored in a `SKDRequestDictionary`.
diff --git a/sourcekit-lsp/Sources/SourceKitD/SKDResponse.swift b/sourcekit-lsp/Sources/SourceKitD/SKDResponse.swift
index 6f7ec1ba..a48af403 100644
--- a/sourcekit-lsp/Sources/SourceKitD/SKDResponse.swift
+++ b/sourcekit-lsp/Sources/SourceKitD/SKDResponse.swift
@@ -19,6 +19,8 @@ import Glibc
 import Musl
 #elseif canImport(CRT)
 import CRT
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 public final class SKDResponse: Sendable {
diff --git a/sourcekit-lsp/Sources/SourceKitD/SKDResponseArray.swift b/sourcekit-lsp/Sources/SourceKitD/SKDResponseArray.swift
index c2d4da58..79c32ee5 100644
--- a/sourcekit-lsp/Sources/SourceKitD/SKDResponseArray.swift
+++ b/sourcekit-lsp/Sources/SourceKitD/SKDResponseArray.swift
@@ -18,6 +18,8 @@ import Glibc
 import Musl
 #elseif canImport(CRT)
 import CRT
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 public final class SKDResponseArray: Sendable {
diff --git a/sourcekit-lsp/Sources/SourceKitD/SKDResponseDictionary.swift b/sourcekit-lsp/Sources/SourceKitD/SKDResponseDictionary.swift
index efb5d761..e8b8d250 100644
--- a/sourcekit-lsp/Sources/SourceKitD/SKDResponseDictionary.swift
+++ b/sourcekit-lsp/Sources/SourceKitD/SKDResponseDictionary.swift
@@ -18,6 +18,8 @@ import Glibc
 import Musl
 #elseif canImport(CRT)
 import CRT
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 public final class SKDResponseDictionary: Sendable {
diff --git a/sourcekit-lsp/Sources/SKSupport/dlopen.swift b/sourcekit-lsp/Sources/SKSupport/dlopen.swift
index 4adf5b44..07573da1 100644
--- a/sourcekit-lsp/Sources/SKSupport/dlopen.swift
+++ b/sourcekit-lsp/Sources/SKSupport/dlopen.swift
@@ -19,6 +19,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Android)
+import Android
 #endif
 
 public final class DLHandle {
diff --git a/swiftpm/Sources/Basics/Environment/Environment.swift b/swiftpm/Sources/Basics/Environment/Environment.swift
index 7d065982bdd..4fe22853e7c 100644
--- a/swiftpm/Sources/Basics/Environment/Environment.swift
+++ b/swiftpm/Sources/Basics/Environment/Environment.swift
@@ -19,6 +19,8 @@ import Musl
 #elseif os(Windows)
 import CRT
 import WinSDK
+#elseif canImport(Bionic)
+import Bionic
 #else
 import Darwin.C
 #endif
diff --git a/swiftpm/Sources/Build/TestObservation.swift b/swiftpm/Sources/Build/TestObservation.swift
index 98a8f5f8e1e..e707910e6f0 100644
--- a/swiftpm/Sources/Build/TestObservation.swift
+++ b/swiftpm/Sources/Build/TestObservation.swift
@@ -132,6 +132,8 @@ public func generateTestObservationCode(buildParameters: BuildParameters) -> Str
         @_exported import WinSDK
         #elseif os(WASI)
         @_exported import WASILibc
+        #elseif canImport(Bionic)
+        @_exported import Bionic
         #else
         @_exported import Darwin.C
         #endif
diff --git a/swiftpm/Sources/CoreCommands/SwiftCommandState.swift b/swiftpm/Sources/CoreCommands/SwiftCommandState.swift
index 4b5bb530025..8a58a060a1d 100644
--- a/swiftpm/Sources/CoreCommands/SwiftCommandState.swift
+++ b/swiftpm/Sources/CoreCommands/SwiftCommandState.swift
@@ -39,6 +39,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 import func TSCBasic.exec
