diff --git a/swift-driver/Sources/SwiftDriver/Driver/ToolExecutionDelegate.swift b/swift-driver/Sources/SwiftDriver/Driver/ToolExecutionDelegate.swift
index 4d6577d0..0eef6e57 100644
--- a/swift-driver/Sources/SwiftDriver/Driver/ToolExecutionDelegate.swift
+++ b/swift-driver/Sources/SwiftDriver/Driver/ToolExecutionDelegate.swift
@@ -19,6 +19,8 @@ import WinSDK
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #else
 #error("Missing libc or equivalent")
 #endif
@@ -136,7 +136,7 @@ import var TSCBasic.stdoutStream
         }
 #else
       case .signalled(let signal):
-        let errorMessage = strsignal(signal).map { String(cString: $0) } ?? ""
+        let errorMessage = String(cString: strsignal(signal))
         messages = constructJobSignalledMessages(job: job, error: errorMessage, output: output,
                                                  signal: signal, pid: pid).map {
           ParsableMessage(name: job.kind.rawValue, kind: .signalled($0))
diff --git a/swift-driver/Sources/SwiftDriver/SwiftScan/Loader.swift b/swift-driver/Sources/SwiftDriver/SwiftScan/Loader.swift
index 92ec883f..2ab4e6fd 100644
--- a/swift-driver/Sources/SwiftDriver/SwiftScan/Loader.swift
+++ b/swift-driver/Sources/SwiftDriver/SwiftScan/Loader.swift
@@ -21,6 +21,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Android)
+import Android
 #endif
 
 internal enum Loader {
diff --git a/swift-driver/Sources/SwiftDriver/Utilities/DateAdditions.swift b/swift-driver/Sources/SwiftDriver/Utilities/DateAdditions.swift
index fcad533d..8d924dbd 100644
--- a/swift-driver/Sources/SwiftDriver/Utilities/DateAdditions.swift
+++ b/swift-driver/Sources/SwiftDriver/Utilities/DateAdditions.swift
@@ -18,6 +18,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 /// Represents a time point value with nanosecond precision.
diff --git a/swift-driver/Sources/SwiftDriver/Utilities/System.swift b/swift-driver/Sources/SwiftDriver/Utilities/System.swift
index c4fdaa89..a3b4d451 100644
--- a/swift-driver/Sources/SwiftDriver/Utilities/System.swift
+++ b/swift-driver/Sources/SwiftDriver/Utilities/System.swift
@@ -16,6 +16,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Android)
+import Android
 #endif
 
 func argumentNeedsQuoting(_ argument: String) -> Bool {
diff --git a/swift-driver/Sources/swift-build-sdk-interfaces/main.swift b/swift-driver/Sources/swift-build-sdk-interfaces/main.swift
index 07eced80..a12a594a 100644
--- a/swift-driver/Sources/swift-build-sdk-interfaces/main.swift
+++ b/swift-driver/Sources/swift-build-sdk-interfaces/main.swift
@@ -19,6 +19,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Bionic)
+import Bionic
 #endif
 
 import class TSCBasic.DiagnosticsEngine
diff --git a/swift-driver/Sources/swift-driver/main.swift b/swift-driver/Sources/swift-driver/main.swift
index 8126b659..e1f2f56f 100644
--- a/swift-driver/Sources/swift-driver/main.swift
+++ b/swift-driver/Sources/swift-driver/main.swift
@@ -20,6 +20,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Android)
+import Android
 #endif
 
 import Dispatch
