diff --git a/sourcekit-lsp/Sources/SourceKitD/dlopen.swift b/sourcekit-lsp/Sources/SourceKitD/dlopen.swift
index 4adf5b44..07573da1 100644
--- a/sourcekit-lsp/Sources/SourceKitD/dlopen.swift
+++ b/sourcekit-lsp/Sources/SourceKitD/dlopen.swift
@@ -19,6 +19,8 @@ import Darwin
 import Glibc
 #elseif canImport(Musl)
 import Musl
+#elseif canImport(Android)
+import Android
 #endif
 
 public final class DLHandle: Sendable {
