diff --git a/yams/Sources/Yams/Representer.swift b/yams/Sources/Yams/Representer.swift
index a749c52..b74ef8e 100644
--- a/yams/Sources/Yams/Representer.swift
+++ b/yams/Sources/Yams/Representer.swift
@@ -14,6 +14,10 @@ private let cpow: (_: Double, _: Double) -> Double = Darwin.pow
 #elseif os(Windows)
 import ucrt
 private let cpow: (_: Double, _: Double) -> Double = ucrt.pow
+#elseif canImport(Android)
+import CoreFoundation
+import Android
+private let cpow: (_: Double, _: Double) -> Double = Android.pow
 #else
 import CoreFoundation
 import Glibc
